### ISSUE TITLE
Token distribution tool fails when using Alchemy provider

### ISSUE THREAD

**Issue by franckc** [OriginProtocol/origin#2432]
On mainnet, the Origin Rewards [token distribution script](https://github.com/OriginProtocol/origin/blob/master/infra/growth/src/scripts/distributeRewards.js) fails when using Alchemy, but it works when using Infura. The first transaction succeeds but then the 2nd one fails with error "Transaction gas price is too low. There is another transaction with same nonce in the queue. Try increasing the gas price or incrementing the nonce."

Here is the [transaction code](https://github.com/OriginProtocol/origin/blob/master/packages/token/src/contractHelper.js#L24).
Looking at the logs, I wonder if with Alchemy, the call to web3.eth.getTransactionReceipt returns even if transaction hasn't been mined yet. Which would explain why the logs show a blockNumber of null for the 1st transaction and that the same nonce was used for the 2nd transaction and caused an error...

Logs with error on 2nd transaction:
2019-06-07T03:50:15.880Z [INFO]  distRewards: Starting rewards distribution job.
2019-06-07T03:50:15.881Z [INFO]  distRewards: Config:
2019-06-07T03:50:15.883Z [INFO]  distRewards: { networkId: 1, doIt: true, gasPriceMultiplier: '1.3' }
Network=1 URL=https://eth-mainnet.alchemyapi.io/jsonrpc/pQyjqDOjFO-ma8XIFDb4P7JF1wI5Hz2a Using private key for account 0x2D00c3c132A0567bbbb45ffcfD8c6543E08FF626
2019-06-07T03:50:16.350Z [INFO]  tokenDistributor: TokenDistributor:
2019-06-07T03:50:16.350Z [INFO]  tokenDistributor:   Network id: 1
2019-06-07T03:50:16.350Z [INFO]  tokenDistributor:   Provider URL: https://eth-mainnet.alchemyapi.io/jsonrpc/pQyjqDOjFO-ma8XIFDb4P7JF1wI5Hz2a
2019-06-07T03:50:16.350Z [INFO]  tokenDistributor:   Address: 0x2D00c3c132A0567bbbb45ffcfD8c6543E08FF626
2019-06-07T03:50:16.351Z [INFO]  tokenDistributor:   Balance: 29544.9799999999999993 OGN
2019-06-07T03:50:16.351Z [INFO]  tokenDistributor:   Gas price multiplier: 1.3
2019-06-07T03:50:16.674Z [INFO]  distRewards: Calculating rewards for campaign 3 (growth.may2019.short_name)
2019-06-07T03:50:16.740Z [INFO]  distRewards: Distribution of 25 to 0x85f161e0d19e13b56e4118f2eff6f3347e7209ce
2019-06-07T03:50:16.751Z [INFO]  distRewards: Checked there is no existing payout row for 0x85f161e0d19e13b56e4118f2eff6f3347e7209ce
2019-06-07T03:50:16.751Z [INFO]  distRewards: Distributing reward of 25 OGN to 0x85f161e0d19e13b56e4118f2eff6f3347e7209ce
2019-06-07T03:50:33.845Z [INFO]  tokenDistributor: Blockchain transaction
2019-06-07T03:50:33.845Z [INFO]  tokenDistributor:   NetworkId:         1
2019-06-07T03:50:33.845Z [INFO]  tokenDistributor:   GasMultiplier:     1.3
2019-06-07T03:50:33.846Z [INFO]  tokenDistributor:   GasPrice:          13000000000
2019-06-07T03:50:33.846Z [INFO]  tokenDistributor:   Amount (natural):  25000000000000000000
2019-06-07T03:50:33.846Z [INFO]  tokenDistributor:   Amount (tokens):   25
2019-06-07T03:50:33.846Z [INFO]  tokenDistributor:   From:              0x2D00c3c132A0567bbbb45ffcfD8c6543E08FF626
2019-06-07T03:50:33.846Z [INFO]  tokenDistributor:   To:                0x85f161e0d19e13b56e4118f2eff6f3347e7209ce
2019-06-07T03:50:33.846Z [INFO]  tokenDistributor:   TxnHash:           0x79f9d6172f16fcc952ae53de8a82f1523f1e7be4f50ed2ebb7e5cb7ae67f62b8
2019-06-07T03:50:33.846Z [INFO]  tokenDistributor:   BlockNumber:       null
2019-06-07T03:50:33.873Z [INFO]  distRewards: Distribution of 125 to 0x4fda7f2b54adb01e62cfb986e2d455dc6447761f
2019-06-07T03:50:33.879Z [INFO]  distRewards: Checked there is no existing payout row for 0x4fda7f2b54adb01e62cfb986e2d455dc6447761f
2019-06-07T03:50:33.879Z [INFO]  distRewards: Distributing reward of 125 OGN to 0x4fda7f2b54adb01e62cfb986e2d455dc6447761f
(node:455) UnhandledPromiseRejectionWarning: Error: Returned error: Transaction gas price is too low. There is another transaction with same nonce in the queue. Try increasing the gas price or incrementing the nonce.
    at Object.ErrorResponse (/app/node_modules/web3-core-helpers/src/errors.js:29:16)
    at /app/node_modules/web3-core-requestmanager/src/index.js:140:36
    at XMLHttpRequest.request.onreadystatechange (/app/node_modules/web3-providers-http/src/index.js:77:13)
    at XMLHttpRequestEventTarget.dispatchEvent (/app/node_modules/web3-providers-http/node_modules/xhr2/lib/xhr2.js:64:18)
    at XMLHttpRequest._setReadyState (/app/node_modules/web3-providers-http/node_modules/xhr2/lib/xhr2.js:354:12)
    at XMLHttpRequest._onHttpResponseEnd (/app/node_modules/web3-providers-http/node_modules/xhr2/lib/xhr2.js:509:12)
    at IncomingMessage.<anonymous> (/app/node_modules/web3-providers-http/node_modules/xhr2/lib/xhr2.js:469:24)
    at IncomingMessage.emit (events.js:203:15)
    at endReadableNT (_stream_readable.js:1129:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)

**Comment by franckc** [comment]
CC @mikeshultz in case you have any idea. Note: this is low priority since the tool works fine with Infura. But an interesting difference to know about...

**Comment by mikeshultz** [comment]
Not really awake yet, but that really shouldn't happen.  You get [verbose output](https://github.com/OriginProtocol/origin/blob/master/packages/token/src/contractHelper.js#L92) by chance?

**Comment by mikeshultz** [comment]
I think this [might be a parity thing](https://wiki.parity.io/JSONRPC-eth-module.html#eth_gettransactionreceipt), which Alchemy is apparently based on.  Here's a note from their docs:

>  **Note** That the receipt is available even for pending transactions.

I don't currently have a Parity node to mess around with, but I could set one up if you want me to confirm this with certainty.

**Comment by mikeshultz** [comment]
And [the Alchemy docs contradict Parity's](https://docs.alchemyapi.io/docs/alchemy-api-reference#section-eth_gettransactionreceipt)...

> Note That the receipt is not available for pending transactions.

**Comment by mikeshultz** [comment]
Alchemy folks confirmed in telegram that that is an error in their docs.  It **WILL** return a receipt for pending transactions... this could have significantly wider impacts across... well, anything waiting on transaction receipts and could be the cause of that confusion reported in #bug-reports on discord.  

Will investigate a bit further, and maybe get a new GH issue going to work on this.  It's probably best to support all JSON-RPC services if we can.

**Comment by franckc** [comment]
Great. Thanks for digging into this. I'll keep this issue open to fix the code in the token package...

**Comment by franckc** [comment]
Should get fixed by #2646
